### ISSUE TITLE
[3.0.0] EVP_PKEY_CTX_set_ec_param_enc() is not supported

### ISSUE THREAD

**Issue by romen** [openssl/openssl#12852]
>> @danbev [said](https://github.com/openssl/openssl/issues/12102#issuecomment-653438152)
>>
>> I'm currently working on updating Node.js to build and run against OpenSSL 3.x and we use EVP_PKEY_CTX_set_ec_param_enc. (I've extracted this into a standalone program, ec.c which might be easier to follow)
>>
>> This is causing test failures for us at the moment, but we can add checks to skip them until a fix lands upstream. I just wanted to let you know that we've run into this, and please let us know if there is anything we should do differently.

> @romen [said](https://github.com/openssl/openssl/issues/12102#issuecomment-653457021)
>
> @danbev with named curve it should be working, what failure do you see?

@romen There is no OpenSSL failure/error reported as far as I can tell, below is the output of the standalone program `ec` and the error in Node.js. 

What I see happening is that the function [legacy_ctrl_to_param](https://github.com/openssl/openssl/blob/94fcec0902ec6df54a6286f61e35fa6f93815fd6/crypto/evp/pmeth_lib.c#L808) is called and since there is no case that matches [EVP_PKEY_EC](https://github.com/openssl/openssl/blob/94fcec0902ec6df54a6286f61e35fa6f93815fd6/crypto/evp/pmeth_lib.c#L850) the [return statement](https://github.com/openssl/openssl/blob/94fcec0902ec6df54a6286f61e35fa6f93815fd6/crypto/evp/pmeth_lib.c#L938) will be reached and 0 will be returned. This is then checked in Node.js as interpreted as an error, see details below:

<details>
<summary>ec output</summary>

```console
$ ./ec 
Elliptic Curve example
curve_nid of secp256k1: 714
EVP_PKEY_CTX_set_ec_param_enc retuned: 0
EVP_PKEY_CTX_set_ec_param_enc failed
errno: 0, error:00000000:lib(0)::reason(0)

```
</details>
<details>
 <summary>Node.js error</summary>

```console
$ ./out/Debug/node test/parallel/test-crypto-keygen.js
./out/Debug/node[876391]: ../src/node_crypto.cc:6350:void node::crypto::GenerateKeyPairJob::ToResult(v8::Local<v8::Value>*, v8::Local<v8::Value>*, v8::Local<v8::Value>*): Assertion `!errors_.empty()' failed.
 1: 0xe4f878 node::DumpBacktrace(_IO_FILE*) [./out/Debug/node]
 2: 0xf07561 node::Abort() [./out/Debug/node]
 3: 0xf07617  [./out/Debug/node]
 4: 0x10d636d node::crypto::GenerateKeyPairJob::ToResult(v8::Local<v8::Value>*, v8::Local<v8::Value>*, v8::Local<v8::Value>*) [./out/Debug/node]
 5: 0x10d620e node::crypto::GenerateKeyPairJob::AfterThreadPoolWork() [./out/Debug/node]
 6: 0x10d40a5 node::crypto::CryptoJob::AfterThreadPoolWork(int) [./out/Debug/node]
 7: 0xec1b1c node::ThreadPoolWork::ScheduleWork()::{lambda(uv_work_s*, int)#2}::operator()(uv_work_s*, int) const [./out/Debug/node]
 8: 0xec1b42 node::ThreadPoolWork::ScheduleWork()::{lambda(uv_work_s*, int)#2}::_FUN(uv_work_s*, int) [./out/Debug/node]
 9: 0x1d6dabc  [./out/Debug/node]
10: 0x1d6da01  [./out/Debug/node]
11: 0x1d72249  [./out/Debug/node]
12: 0x1d898a8  [./out/Debug/node]
13: 0x1d72bca uv_run [./out/Debug/node]
14: 0xf752a1 node::NodeMainInstance::Run() [./out/Debug/node]
15: 0xeb4e45 node::Start(int, char**) [./out/Debug/node]
16: 0x23331b2 main [./out/Debug/node]
17: 0x7f319142d1a3 __libc_start_main [/lib64/libc.so.6]
18: 0xdfecee _start [./out/Debug/node]
Aborted (core dumped)
```
</details>

**Comment by romen** [comment]
While #12604 added most of the provider support for `explicit parameters`/`named curve` encoding, `EVP_PKEY_CTX_ctrl()` is missing support for `EVP_PKEY_CTRL_EC_PARAM_ENC`.

It is available in the legacy codepath:

https://github.com/openssl/openssl/blob/9f604ca13ddc99e17ba37fed9281fbd1b71149a9/crypto/ec/ec_pmeth.c#L245-L251

But a provider-native `ctx` goes through `legacy_ctrl_to_param()`, where support for this control is missing:

https://github.com/openssl/openssl/blob/9f604ca13ddc99e17ba37fed9281fbd1b71149a9/crypto/evp/pmeth_lib.c#L1249-L1281

Evidently we are also missing tests for this feature!

Also notice that in 1.1.1 `pkey_ec_ctrl()` was returning `-2` for unsupported controls, while `legacy_ctrl_to_param()` is returning `0` in this case.

**Comment by romen** [comment]
@slontis do you have spare cycles to look into this?

I don't think fixing this will require changes to the API so maybe there is no need to rush and fix this before beta1.

**Comment by romen** [comment]
@danbev of course if you wanted to try and fix this on your own, in a way similar to what is done for the other `cmd` in `legacy_ctrl_to_param()` such a PR would be more than welcome!

**Comment by slontis** [comment]
I will do it in about 6 hours time (when I am supposed to wake up :) ).. Let me know if you try to do it before then @danbev.

**Comment by levitte** [comment]
Mind if I play?

**Comment by levitte** [comment]
Tentative fix in #12853, but it's lacking a test.  Help appreciated.
